torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1325,7 +1383,15 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                 "FSDP only works with gradients that don't require gradients"
             )
 
-        self._free_full_params(cast(List[FlatParameter], [param]))
+        if self._require_backward_grad_sync or self.reshard_after_forward:
+            # Free full params. As a special case, we don't free the full params
+            # when in a ``no_sync`` context (as inversely indicated by
+            # ``self._require_backward_grad_sync``) and
+            # self.reshard_after_forward = False (sharding_strategy=SHARD_GRAD_OP),
+            # since the params will not get updated before the next forward,
+            # there is no need to call all_gather to sync params among ranks. This
+            # saves networking bandwidth but uses more GPU memory.
```

Comment:
A bit confused on this comment - 

regardless of the setting of reshard_after_forward, the param will not get updated before the next forward when in a no_sync manager right? Could we clarify that? 

Also, are we sure that we want this as the default, i.e. the GPU/networking bandwidth trade off is the correct one to make? In general I think we may want to carefully track these tradeoffs that we hardcode in FSDP and eventually expose knobs to allow users to tune (or run experiments and tune / offer recommendations ourselves). What do you think?

Reply:
sounds good, let me clarify the comments. "reshard_after_forward" is kind of tune knob here, if it is full_shard, no_sync model will do more network communication but save some memory, if it is grad_op, no_sync model will use more memory but save network overhead. let me clarify that in the comment as well